Solves ctrl+click on functions by ignoring the cursor

2shady4u opened the pull request:
(This is a new branch containing the same changes as in PR #31112 as something went wrong with the forks...)

Solves #31036, #20393 and issues discussed in #7080 (partially)

I went through the dismal spaghetti and conquered the legacy.
This PR solves an issue with blocking behavior of TK_CURSOR when control+clicking on an identifier (to jump to the function definition).

**Solution at work:**
(This is not possible with the current gdscript editor!)
![click](https://user-images.githubusercontent.com/42484461/64069865-8e1b0c80-cc53-11e9-9665-139a5e7519f3.gif)

**What does it solve?**
Control+clicking on a function call should normally jump to wherever the function definition is placed. This jump did not happen when the function definition was placed after the function call (see #31036).

**What was the problem?**
The parser expects the next token to be an open parenthesis and the current one to be a literal for proper parsing purposes. Now when the cursor is also part of the token array, this cursor blocks and exits the parser. The result is that ONLY the functions that were defined BEFORE the cursor would be added to the functions array of the Parser. As a result... finding functions that were defined AFTER the cursor was impossible.

**Is this the best implementation?**
I think there might be a more elegant solution somewhere 🤔 (or not)
Please guide me into the right direction!

Cheerios!

bojidar-bg commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -238,6 +238,7 @@ GDScriptParser::Node *GDScriptParser::_parse_expression(Node *p_parent, bool p_s
 	Node *expr = NULL;
 
 	int op_line = tokenizer->get_token_line(); // when operators are created at the bottom, the line might have been changed (\n found)
+	int next_valid_offset;
```
Does this really need to be outside the loop?

2shady4u replied:
I guess not 🤔 
Would you rather have this only when it is used?

bojidar-bg replied:
Yes, since it clears up the logic; if the compiler feels like moving it out, it will.

2shady4u replied:
Ok, I changed it.

2shady4u committed:
Solves ctrl+click on functions by ignoring the cursor


Obeyed CLANG format rules


Obeying CLANG format rules attempt 2


Obeying CLANG format rules attempt 3


Clean up


Fixed runaway while loop


Removed int initialization

akien-mga commented:
Thanks!
